### PR TITLE
[WIP] Significant upgrade of non-linear solver

### PR DESCRIPTION
**Pull request by blaisb**
Reduce number of RHS evaluation in the non-linear solver and add a few options which enable more flexibility.
Add a inexact jacobian newton solver that only renews the jacobian matrix when the solution is not progressing fast enough.
Still need to add a test so this is a WIP, but @Luckabarbeau and @lpsaavedra can you take a look at it?

### CODE REVIEW

File: include/core/inexact_newton_iteration_non_linear_solver.h [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+          if (current_res < this->params.step_tolerance * last_res ||
+              last_res < this->params.tolerance)
+            {
+              if (current_res > this->params.matrix_tolerance * last_res)
+                matrix_requires_assembly = true;
+              else
+                matrix_requires_assembly = false;
+
+              break;
+            }
```
Comment (Luckabarbeau) [inline include/core/inexact_newton_iteration_non_linear_solver.h:173]: Maybe add a small comment on how this section works.

File: source/solvers/gls_sharp_navier_stokes.cc [hunk 2]
```diff
@@ -1626,6 +1625,7 @@ GLSSharpNavierStokesSolver<dim>::sharp_edge()
                           support_points[local_dof_indices[i]]);
                       typename DoFHandler<dim>::active_cell_iterator cell_2;
                       bool particle_close_to_wall = false;
+                      (void)particle_close_to_wall;
```
Comment (Luckabarbeau) [inline source/solvers/gls_sharp_navier_stokes.cc:1628]: Not sure why a void is needed here.

**Review by Luckabarbeau — APPROVED** [review]
Very good PR  it will save a lot of useless calculation of the RHS.

File: include/core/parameters.h [hunk 3]
```diff
@@ -381,15 +382,33 @@ namespace Parameters
...
+
 
     // Maximal number of iterations for the Newton solver
     unsigned int max_iterations;
 
     // Residual precision
     unsigned int display_precision;
 
+
+    // Force RHS recalculation at the beggining of every non-linear steps
```
Comment (lpsaavedra) [inline include/core/parameters.h:394]:
```suggestion
    // Force RHS recalculation at the beginning of every non-linear steps
```

File: include/core/parameters.h [hunk 4]
```diff
@@ -381,15 +382,33 @@ namespace Parameters
...
+    // This is required if there is a fixed point component to the non-linear
+    // solver that is changed at the beggining of every newton iteration.
```
Comment (lpsaavedra) [inline include/core/parameters.h:396]:
```suggestion
    // solver that is changed at the beginning of every newton iteration.
```

**Review by lpsaavedra — APPROVED** [review]
Nice addition to the non-linear solvers and good solution as KINSOL is not working as it should at the moment. Found some typos in the documentation, but other than that it seems good. I also liked the distinction between current_res and global_res, it makes everything clearer.

**Pull request merged**